#include "homepage.h"
#include "ui_homepage.h"
#include "datamanager.h"
#include "scheduledialog.h"
#include <QDate>
#include <QMessageBox>

HomePage::HomePage(int userId, QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::HomePage)
    , m_userId(userId)
{
    ui->setupUi(this);

    // ì‹œê·¸ë„-ìŠ¬ë¡¯ ì—°ê²°
    connect(ui->scheduleList, &QListWidget::itemClicked,
            this, &HomePage::onScheduleItemClicked);
    connect(ui->editButton, &QPushButton::clicked,
            this, &HomePage::onEditButtonClicked);
    connect(ui->deleteButton, &QPushButton::clicked,
            this, &HomePage::onDeleteButtonClicked);

    loadAISummary();
    loadUpcomingSchedules();
}

HomePage::~HomePage()
{
    delete ui;
}

void HomePage::loadAISummary()
{
    // TODO: AIServiceë¥¼ í†µí•´ ì‹¤ì œ AI ìš”ì•½ ê°€ì ¸ì˜¤ê¸°
    QString today = QDate::currentDate().toString("yyyy-MM-dd");
    QString summary = DataManager::instance().getCachedSummary(m_userId, today);

    if (summary.isEmpty()) {
        ui->aiSummaryText->setText("AI ìš”ì•½ì„ ë¶ˆëŸ¬ì˜¤ëŠ” ì¤‘...");
        // TODO: AIServiceë¡œ ìš”ì•½ ìƒì„±
    } else {
        ui->aiSummaryText->setText(summary);
    }
}

void HomePage::refreshSchedules()
{
    loadUpcomingSchedules();
    loadAISummary();
}

void HomePage::loadUpcomingSchedules()
{
    // ê¸°ì¡´ ë°ì´í„° í´ë¦¬ì–´
    ui->scheduleList->clear();
    m_itemToScheduleId.clear();

    // ì˜¤ëŠ˜ë¶€í„° 7ì¼ê°„ì˜ ì¼ì • ê°€ì ¸ì˜¤ê¸°
    QList<QVariantMap> schedules = DataManager::instance().getSchedulesForNextDays(m_userId, 7);

    if (schedules.isEmpty()) {
        QListWidgetItem *item = new QListWidgetItem("ì¼ì •ì´ ì—†ìŠµë‹ˆë‹¤.");
        item->setFlags(item->flags() & ~Qt::ItemIsSelectable);  // ì„ íƒ ë¶ˆê°€
        ui->scheduleList->addItem(item);
        return;
    }

    // ì¼ì •ì„ ë¦¬ìŠ¤íŠ¸ì— ì¶”ê°€
    for (const QVariantMap &schedule : schedules) {
        QString date = schedule["date"].toString();
        QString startTime = schedule["startTime"].toString();
        QString title = schedule["title"].toString();
        int scheduleId = schedule["id"].toInt();

        // ë‚ ì§œ í˜•ì‹ ë³€í™˜: 2024-11-20 -> 11/20 (ìˆ˜)
        QDate qDate = QDate::fromString(date, "yyyy-MM-dd");
        QString dayOfWeek = QLocale(QLocale::Korean).dayName(qDate.dayOfWeek(), QLocale::ShortFormat);
        QString formattedDate = qDate.toString("MM/dd");

        // ì‹œê°„ í˜•ì‹: HH:mm:ss -> HH:mm
        QString formattedTime = startTime.left(5);

        // ë¦¬ìŠ¤íŠ¸ ì•„ì´í…œ í…ìŠ¤íŠ¸ ìƒì„±
        QString itemText = QString("ğŸ“Œ %1 (%2) %3 - %4")
                               .arg(formattedDate)
                               .arg(dayOfWeek)
                               .arg(formattedTime)
                               .arg(title);

        QListWidgetItem *item = new QListWidgetItem(itemText);
        ui->scheduleList->addItem(item);

        // ì•„ì´í…œê³¼ ìŠ¤ì¼€ì¤„ ID ë§¤í•‘ ì €ì¥
        m_itemToScheduleId[item] = scheduleId;
    }
}

void HomePage::onScheduleItemClicked(QListWidgetItem *item)
{
    // ì„ íƒ ë¶ˆê°€ëŠ¥í•œ ì•„ì´í…œì¸ ê²½ìš° ë¬´ì‹œ
    if (!(item->flags() & Qt::ItemIsSelectable)) {
        return;
    }

    // ìŠ¤ì¼€ì¤„ ID ê°€ì ¸ì˜¤ê¸°
    int scheduleId = m_itemToScheduleId.value(item, -1);
    if (scheduleId == -1) {
        return;
    }

    // ìƒì„¸ ì •ë³´ í‘œì‹œ
    displayScheduleDetail(scheduleId);

    // ë²„íŠ¼ í™œì„±í™”
    ui->editButton->setEnabled(true);
    ui->deleteButton->setEnabled(true);
}

void HomePage::displayScheduleDetail(int scheduleId)
{
    QVariantMap schedule = DataManager::instance().getScheduleById(scheduleId);

    if (schedule.isEmpty()) {
        ui->scheduleDetail->setText("ì¼ì • ì •ë³´ë¥¼ ë¶ˆëŸ¬ì˜¬ ìˆ˜ ì—†ìŠµë‹ˆë‹¤.");
        return;
    }

    // ìƒì„¸ ì •ë³´ HTML í˜•ì‹ìœ¼ë¡œ í‘œì‹œ
    QString html = QString(
                       "<h3 style='color: #2196F3; margin-bottom: 10px;'>%1</h3>"
                       "<p style='margin: 5px 0;'><b>ğŸ“… ë‚ ì§œ:</b> %2</p>"
                       "<p style='margin: 5px 0;'><b>ğŸ• ì‹œê°„:</b> %3 ~ %4</p>"
                       "<p style='margin: 5px 0;'><b>ğŸ“ ì¥ì†Œ:</b> %5</p>"
                       "<p style='margin: 5px 0;'><b>ğŸ·ï¸ ì¹´í…Œê³ ë¦¬:</b> %6</p>"
                       "<hr style='margin: 10px 0; border: none; border-top: 1px solid #ddd;'>"
                       "<p style='margin: 5px 0;'><b>ğŸ“ ë©”ëª¨:</b></p>"
                       "<p style='margin: 5px 0; padding-left: 10px;'>%7</p>"
                       ).arg(schedule["title"].toString())
                       .arg(schedule["date"].toString())
                       .arg(schedule["startTime"].toString().left(5))
                       .arg(schedule["endTime"].toString().left(5))
                       .arg(schedule["location"].toString().isEmpty() ? "ì—†ìŒ" : schedule["location"].toString())
                       .arg(schedule["category"].toString().isEmpty() ? "ì—†ìŒ" : schedule["category"].toString())
                       .arg(schedule["memo"].toString().isEmpty() ? "ì—†ìŒ" : schedule["memo"].toString());

    ui->scheduleDetail->setHtml(html);
}

void HomePage::onEditButtonClicked()
{
    QListWidgetItem *currentItem = ui->scheduleList->currentItem();
    if (!currentItem) return;

    int scheduleId = m_itemToScheduleId.value(currentItem, -1);
    if (scheduleId == -1) return;

    // ì¼ì • ì •ë³´ ê°€ì ¸ì˜¤ê¸°
    QVariantMap scheduleData = DataManager::instance().getScheduleById(scheduleId);

    // ScheduleDialog ì—´ê¸° (ìˆ˜ì • ëª¨ë“œ)
    ScheduleDialog dialog(scheduleData, this);

    if (dialog.exec() == QDialog::Accepted) {
        QVariantMap data = dialog.getScheduleData();

        bool success = DataManager::instance().updateSchedule(
            scheduleId,
            data["title"].toString(),
            data["date"].toString(),
            data["start_time"].toString(),
            data["end_time"].toString(),
            data["location"].toString(),
            data["memo"].toString(),
            data["category"].toString()
            );

        if (success) {
            // DataManagerê°€ scheduleChanged ì‹œê·¸ë„ ë°œìƒ
            // MainWindowê°€ ìë™ìœ¼ë¡œ HomePage ê°±ì‹ í•´ì¤Œ
            QMessageBox::information(this, "ìˆ˜ì • ì™„ë£Œ", "ì¼ì •ì´ ìˆ˜ì •ë˜ì—ˆìŠµë‹ˆë‹¤.");
        } else {
            QMessageBox::warning(this, "ì˜¤ë¥˜", "ì¼ì • ìˆ˜ì •ì— ì‹¤íŒ¨í–ˆìŠµë‹ˆë‹¤.");
        }
    }
}

void HomePage::onDeleteButtonClicked()
{
    QListWidgetItem *currentItem = ui->scheduleList->currentItem();
    if (!currentItem) {
        return;
    }

    int scheduleId = m_itemToScheduleId.value(currentItem, -1);
    if (scheduleId == -1) {
        return;
    }

    // ì‚­ì œ í™•ì¸
    QMessageBox::StandardButton reply;
    reply = QMessageBox::question(this, "ì¼ì • ì‚­ì œ",
                                  "ì„ íƒí•œ ì¼ì •ì„ ì‚­ì œí•˜ì‹œê² ìŠµë‹ˆê¹Œ?",
                                  QMessageBox::Yes | QMessageBox::No);

    if (reply == QMessageBox::Yes) {
        if (DataManager::instance().deleteSchedule(scheduleId)) {
            QMessageBox::information(this, "ì‚­ì œ ì™„ë£Œ", "ì¼ì •ì´ ì‚­ì œë˜ì—ˆìŠµë‹ˆë‹¤.");

            // ë¦¬ìŠ¤íŠ¸ ìƒˆë¡œê³ ì¹¨
            loadUpcomingSchedules();

            // ìƒì„¸ ì •ë³´ ì´ˆê¸°í™”
            ui->scheduleDetail->setText("ì¼ì •ì„ ì„ íƒí•˜ì„¸ìš”.");
            ui->editButton->setEnabled(false);
            ui->deleteButton->setEnabled(false);
        } else {
            QMessageBox::warning(this, "ì‚­ì œ ì‹¤íŒ¨", "ì¼ì • ì‚­ì œì— ì‹¤íŒ¨í–ˆìŠµë‹ˆë‹¤.");
        }
    }
}
